Fix loading scene for every request on script file for workspace completion

dekaravanhoc opened the pull request:
For workspace completion the Scene for the script is loaded for every completion request and freed afterwards. Resulting in performance issues for scenes including large resources.

Loaded Scenes are now cached for each script. Not loading them again and only freeing them when the connection is terminated or the language server stopped.

This PR fixes #100934

---
I have no experience in C++ development - feedback is of course more than welcome. 
Unit Tests are green and auto formatting should have worked too.

_Bugsquad edit (keywords for easier searching): LSP, language server_

dekaravanhoc committed:
Fix loading scene for every request on script file for workspace completion

For workspace completion the Scene for the script was loaded for every
completion request and freed afterwards. Resulting in performance issues
for scenes including large resources.

Loaded Scenes are now cached for each script and loaded when opening the
script. Not loading them again and only freeing them when the connection
is terminated, the language server stopped or the script is closed.

Add use of a thread for resource loading in background, but loads
resource if its directly requested

Only one load task will be running at a time.

Remove unused header

Add a struct for handling of the loadqueue

Add LSP debug print macro

HolonProduction commented:
This should remove the delay for subsequent completion requests, the delay would still be there for the initial request. I think we could improve on this by using the [textDocument/didOpen](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen) and [textDocument/didClose](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didClose) signals to manage the cache. This way Godot could start loading the owner, when the script gets opened. Depending on the time between opening and typing something (which can be quite a bit in bigger projects from my experience) this might completely remove the delay or decrease it a bit.

Also Godot resource loading system has a builtin cache which would definitely come in handy with the preloading approach. As long as the loaded resource isn't freed it will be reused in the future. We could just cache the references to the loaded scene and re instantiate it since this shouldn't cause much overhead (this would also allow replacing `SceneCache` with a normal `HashMap`).

This could work kinda like this:
- in `didOpen`:
    - use `ResourceLoader.load_threaded_request` to start loading the owner
    - add an invalid `Ref` to the hashmap for the path (we need to keep track of this in case the user opens files but does never complete anything)
- in `completion`:
    - use `ResourceLoader.load_threaded_request` again to make sure a loading process is going on for completions after the first one
    - obtain the resource with `ResourceLoader.load_threaded_get` (blocking like before if the resource isn't fully loaded yet)
    - then safe the resource into the hash map under the path (this will keep the resource in memory thus automatically using the resource loader cache in the future)
- in `didClose` (and on disconnect for external editor crashes):
    - remove entry from hash map and if entry was invalid use `ResourceLoader.load_threaded_get` to prevent leaking results

dekaravanhoc commented:
Hey thanks,

the SceneCache is only a wrapper for a HashMap with a Scene specific clear() function. 

Also as these Scenes are loaded resources and their sub resources also, the ResourceLoader already uses the cache for all of them, if there are used in another loaded Scene.

For the rest, sounds reasonable. I'll look into it. Also I have really fast completion times even at the beginning now with a production build and large resources (~1GB loaded), as the main slowdown came from the queue of loading->freeing->loading->freeing.

HolonProduction commented:
> Also as these Scenes are loaded resources and their sub resources also, the ResourceLoader already uses the cache for all of them, if there are used in another loaded Scene.

Yeah right, my main reason for suggesting only relying on the resource loader cache was the idea with the loading headstart, which might get harder to implement when caching the nodes.

> For the rest, sounds reasonable. I'll look into it. Also I have really fast completion times even at the beginning now with a production build and large resources (~1GB loaded), as the main slowdown came from the queue of loading->freeing->loading->freeing.

That's interesting, I still get the same slow down on the first completion request, with this PR and my MRP (no production build). Maybe we are hitting different bottle necks with loading :thinking:

dekaravanhoc commented:
> That's interesting, I still get the same slow down on the first completion request, with this PR and my MRP (no production build). Maybe we are hitting different bottle necks with loading 🤔

Ah k. I used a production build (ucrt). On my dev build it was in deed still very slow, even for the first input. But I did not stop time, so could just be my imagination over being happy that it worked. 🙃
That said, I think, its a good idea to load the scene when opening the script.

dekaravanhoc commented:
I moved all logic for the Scene loading into its own class SceneCache inside language_server_protocol.

Threaded resource Loading is added and used, when "use thread" is activated for the language server.
Its quite nice, as the loading now does not block the server for auto completion etc. And as far as I can see, the Scene is only used for Node path completion - definitively a win on scenes with large resources.

Using threaded resource loading is a bit of a hassle, as there is no option to kill a specific running task. (Or at least I did not found one.) So the server needs to wait for a running task and clean it afterwards, if the client disconnects or closes the document.
Also running multiple tasks in parallel that might load the same sub resources can bring up errors in the editor, so I opted for one task at a time.

But for now, I think it works quite well.

HolonProduction reviewed (commented):
In general I think that when the completion request arrives and the scene isn't loaded yet, we should not continue without owner and instead wait for loading to finish. The owner is not only used for node path completion, but also for argument options (e.g. suggesting animation names when doing `$AnimationPlayer.play()`).

Otherwise two subsequent completion requests could lead to different results, I think that's not really great for the user experience.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+
+	is_loading = false;
+
+	Ref<Resource> owner_res = ResourceLoader::load_threaded_get(owner_path);
+	if (Object::cast_to<PackedScene>(owner_res.ptr())) {
+		Ref<PackedScene> owner_packed_scene = Ref<PackedScene>(Object::cast_to<PackedScene>(*owner_res));
+		cache[check_path] = owner_packed_scene->instantiate();
```
I have basically no clue of c++ and its behavior. But to me it does not look like as it will work?
load_threaded_get() could return any Ref<Resource> and owner_res would be valid resulting in an error for resources without instantiate()?

HolonProduction replied:
The important part is the templated type. `Ref` contains [implicit conversion](https://github.com/godotengine/godot/blob/master/core/object/ref_counted.h#L133) when assigning different types. So when assigning let's say a `Ref<Script>` to a `Ref<PackedScene>` the resulting ref isn't valid anymore.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -47,6 +50,35 @@
+	HashMap<String, Node *> cache;
+	Ref<GDScriptWorkspace> workspace;
+	bool is_loading = false;
+	HashMap<String, List<String>> owners_path_cache;
```
As there could be multiple possible owner and we load only one resource at a time. We need to know which possible owners where already resolved.
And I think if there is no valid reason to not cache the result of a file system operation inside the scope of a limited system, one should do it. But thats more of a base discussion. It is also needed for the correct behavior, so I think it is ok to use.

HolonProduction replied:
AFAIK editor file system already keeps a representation of the FS in memory so there would be no file system operations in play. But I'll ask again to make sure.

My main concern was about outdated content, but seeing as this is only short term and invalidates as soon as valid owner is loaded it should be fine I guess.

dekaravanhoc replied:
Ah cool. Thanks for clarification.
But yeah, as it is also needed to work through the owners one by one in the limited scope, should be reasonable. 
Initially I had the owners loaded all parallel, but I got errors when loading overlapping resources in parallel. :/

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	scene_cache.workspace = workspace;
+}
+
+void SceneCache::_get_owners(EditorFileSystemDirectory *efsd, String p_path, List<String> &owners) {
```
A thanks. :)
What is the wanted process? Pushing single commits for resolving single suggestions. Squashing them in the end, when all is fine? Or just push all changes in one already squashed commit?
Or something else?

HolonProduction replied:
In general there is a one commit per PR policy. If it helps you, you can work with multiple but you will be asked to squash before the PR gets merged.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -136,6 +139,7 @@ Error GDScriptLanguageProtocol::on_client_connected() {
 
 void GDScriptLanguageProtocol::on_client_disconnected(const int &p_client_id) {
 	clients.erase(p_client_id);
+	scene_cache.clear();
```
Thats right. :)

Initially I thought about a per client cache, but did not want to over engineer on a solution, that might not be good. But as the general concept / code seems to be fine, I can add it. Would probably be the cleanest solution?

HolonProduction replied:
Fair point :+1:  Didn't think about this at all.

dekaravanhoc commented:
Hey, thank you very much for reviewing. :)

Ah ok, I did not look to deep into the usage of the scene itself for auto completion. I get were you are coming from, but I think it is the better user experience to instantly get an auto completion from the language server after opening a script.
Completion suggestions change in both cases. If you wait it changes basically every time from only editor suggestions to editor suggestion + ls. If you do not wait it only changes for the specific completions. 
I think the coolest thing, about the threaded resource load is, that it does not block any other language server operations.

dekaravanhoc commented:
Hey,

so from my side it is finished. Worked the last weeks with this in 4.3 and in master and got no problems.
Let me know if something is off.

Best regards :)

akien-mga commented:
Just a quick update, this still needs a maintainer approval, but as a side note:

Your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

dekaravanhoc commented:
> Just a quick update, this still needs a maintainer approval, but as a side note:
> 
> Your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

Thanks for the heads up. I added the e-mail adress. Should be working now.
Also pushed the latest rebased state.

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	bool remote_use_thread = (bool)_EDITOR_GET("network/language_server/use_thread");
+	if (remote_use_thread && resource_request_queue.has(p_path)) {
+		if (is_loading && resource_request_queue.front() == p_path) {
+			while (is_loading) {
+				_check_thread_for_cache_update();
+				OS::get_singleton()->delay_usec(50000);
+			}
```
`ResourceLoader::load_threaded_get` blocks till the resource finished loading. We should use it here instead of this custom "spin lock".

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	if (!cache.has(p_path)) {
+		return;
+	}
+	if (cache[p_path]) {
+		memdelete(cache[p_path]);
+	}
```
```suggestion
	memdelete_notnull(cache[p_path]);
```

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+
+void SceneCache::clear() {
+	if (is_loading) {
+		while (is_loading) {
+			_check_thread_for_cache_update();
+			OS::get_singleton()->delay_usec(100);
+		}
```
ResourceLoader::load_threaded_get blocks till the resource finished loading. We should use it here instead of this custom "spin lock".

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	}
+	resource_request_queue.clear();
+	for (const KeyValue<String, Node *> &E : cache) {
+		if (E.value) {
+			memdelete(E.value);
+		}
```
```suggestion
		memdelete_notnull(E.value);
```

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -47,6 +50,35 @@
+	bool has(const String &p_path);
+	Node *get(const String &p_path);
+	Node *get_for_uri(const String &p_uri);
+	void set(const String &p_path);
+	void set_for_uri(const String &p_uri);
```
Please rename those to something like `request` or `enqueue`, to make clear that this does not instantly set the value.

HolonProduction commented on modules/gdscript/language_server/gdscript_text_document.cpp:
```diff
@@ -82,6 +83,7 @@ void GDScriptTextDocument::didChange(const Variant &p_param) {
 		doc.text = evt.text;
 	}
 	sync_script_content(doc.uri, doc.text);
+	GDScriptLanguageProtocol::get_singleton()->get_scene_cache()->get_for_uri(doc.uri);
```
The return value is discarded. Why do we need this here?

dekaravanhoc replied:
That was a relic I think, from one of the earliest implementations.

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+ * Does only one threaded request to the ResourceLoader at a time.
+ * Because loading the same subresources in parallel can bring up errors in the editor.
+ * */
+void SceneCache::_add_owner_scene_request(String p_path = "") {
```
This method is doing too much (otherwise you wouldn't need a private default value).

We should split this up in one method for enqueuing new paths. And one for updating the queue. (The former can obviously call to the later)

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	}
+}
+
+void SceneCache::_check_thread_for_cache_update() {
```
Splitting this function up would make sense. The part beginning from `ResourceLoader::load_threaded_get` should be its own function. This would allow to easily remove the custom spin locks from `erase` and `clear`.

HolonProduction reviewed (changes requested):
Sorry for the long delay, had some exams to write.

I tested the PR locally and it seems to work fine.

I have to come back to one thing I brought up before though:

> In general I think that when the completion request arrives and the scene isn't loaded yet, we should not continue without owner and instead wait for loading to finish.

I was going to say, that this is not a deal breaker for me, but I read a bit more into the spec due to some other PR and noticed that implementation of file management (`didOpen`/`didClose`) is not required. I don't know any editor that does not implement those, but since it is part of the spec we should have a contingency plan for it. The easiest one, would be to block the current thread and finish loading inside of `get` (this would only happen on the first request, and not at all if `didOpen` is implemented and finished before).

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+	scene_cache.workspace = workspace;
+}
+
+void SceneCache::_get_owners(EditorFileSystemDirectory *p_efsd, const String &p_path, List<String> &r_owners) {
```
```suggestion
void SceneCache::_get_owners(EditorFileSystemDirectory *p_efsd, const String &p_path, LocalVector<String> &r_owners) {
```
`List` should not be used in new Code anymore as per [the c++ usage guidelines](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html)

HolonProduction commented:
Instead of marking all usages: `List` should not be used in new Code anymore as per [the c++ usage guidelines](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html). The best way to do this here, would probably to use a `LocalVector` instead.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+		}
+	}
+
+	cache[p_path] = owner_scene_node;
```
Here was some small error in the suggestion.

HolonProduction replied:
I guess you mean that `owner_scene_node` should be replaced with `owner_packed_scene->instantiate()`? Or did I miss something else? Only other difference I see is the behavior is nothing is found, but whether a `nullptr` gets inserted or it is just not part of the map didn't matter IIRC.

dekaravanhoc commented:
> Sorry for the long delay, had some exams to write.
> 
> I tested the PR locally and it seems to work fine.
> 
> I have to come back to one thing I brought up before though:
> 
> > In general I think that when the completion request arrives and the scene isn't loaded yet, we should not continue without owner and instead wait for loading to finish.
> 
> I was going to say, that this is not a deal breaker for me, but I read a bit more into the spec due to some other PR and noticed that implementation of file management (`didOpen`/`didClose`) is not required. I don't know any editor that does not implement those, but since it is part of the spec we should have a contingency plan for it. The easiest one, would be to block the current thread and finish loading inside of `get` (this would only happen on the first request, and not at all if `didOpen` is implemented and finished before).

Hey,

maybe I do not understand exactly what you mean by that, but it states quite clear to me that didOpen and didClose are mandatory.
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_synchronization

dekaravanhoc commented:
> Instead of marking all usages: `List` should not be used in new Code anymore as per [the c++ usage guidelines](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html). The best way to do this here, would probably to use a `LocalVector` instead.

Changed it.

HolonProduction commented:
> Hey,
> 
> maybe I do not understand exactly what you mean by that, but it states quite clear to me that didOpen and didClose are mandatory. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_synchronization

Yes, I didn't describe my problem correctly. What I'm talking about is the following part of the spec:

> [Note that a server’s ability to fulfill requests is independent of whether a text document is open or closed.](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen)

So yes clients must implement those notifications, but they are not required to actually open the file before performing other requests. (They do in practice so it's a good optimization to make, but completely relying on this behavior kinda goes against the spec from my understanding)

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -345,4 +351,209 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+
+	String path = resource_request_queue.front();
+	if (!owners_path_cache.has(p_path)) {
+		_get_owners(EditorFileSystem::get_singleton()->get_filesystem(), path, owners_path_cache[p_path]);
```
When assigning you are expecting it to do that. Reading the original code requires knowledge about this quirk. So it's mainly a readability thing for now.

dekaravanhoc commented:
> > Hey,
> > maybe I do not understand exactly what you mean by that, but it states quite clear to me that didOpen and didClose are mandatory. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_synchronization
> 
> Yes, I didn't describe my problem correctly. What I'm talking about is the following part of the spec:
> 
> > [Note that a server’s ability to fulfill requests is independent of whether a text document is open or closed.](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen)
> 
> So yes clients must implement those notifications, but they are not required to actually open the file before performing other requests. (They do in practice so it's a good optimization to make, but completely relying on this behavior kinda goes against the spec from my understanding)

From the Specs:

> The document change notification is sent from the client to the server to signal changes to a text document. Before a client can change a text document it must claim ownership of its content using the [textDocument/didOpen](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didOpen) notification. In 2.0 the shape of the params has changed to include proper version numbers.
> 
> Before requesting information from the server (e.g., [textDocument/completion](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion) or [textDocument/signatureHelp](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_signatureHelp)), the client must ensure that the document’s state is synchronized with the server to guarantee reliable results.

As I understand it, the Client must claim the document with `didOpen` and can not expect reliable results from `completion` without doing so and even needs sending a `didChange` beforehand.
So the question stands from the beginning. Should a completion request be delayed until the requested document path is cached. From a UX side I still think no, but I understand how this is the expected behavior from a technical side.
So I ll add: When the cache for a path is requested, it is finished loading first. 
With the caching started as soon as a document opens, it is still well optimized as you stated.

HolonProduction commented:
Well from how I understand it, if a client did not claim ownership of a file with `didOpen` it is synced by definition, because the state on disc has to be the current state in the client as well.

I still think it's worth to account for it. Because while such a scenario would not be realistic for completion, it might be for other requests (e.g. symbol lookup) which could happen in some sort of "read only" mode. This way we keep the option to introduce owner based behavior in such endpoints. (If we don't account for it, we need to document clearly in which scenarios the owner cache can be used.)

dekaravanhoc commented:
So I added the blocking load, when the resources for a script are not yet loaded.
I also looked into the code again and added console prints for a dev build of the editor. Not quite shure, if thats the way to do it.
Using only breakpoints and the debugger was quite tedious for this. :-) 
Also made some small fixes.
This should be working quite well now.
(Also I was working with the prior state on a custom build for weeks now. Seems to be very stable.)

HolonProduction commented:
> I also looked into the code again and added console prints for a dev build of the editor. Not quite shure, if thats the way to do it.

I think that's what `print_verbose` is for (will only print when Godot runs with the --verbose, flag).

dekaravanhoc commented:
> I think that's what `print_verbose` is for (will only print when Godot runs with the --verbose, flag).

Hmm, ok. But this is also for the production build of the engine, if I understand correctly? My Intention was to only print these messages, when a dev build of the engine is run. I did not want to add this logging in the production build.
I can also remove them or thin them out and use print_verbose.

HolonProduction commented:
Ok, what most other modules do in this case, is defining an own macro for debugging the specific class. [Here are some examples.](https://github.com/search?q=repo%3Agodotengine%2Fgodot%20%2F%2F%23define&type=code) I don't think we need an own macro for the cache, so something like `DEBUG_LSP` would be best for this I guess.

dekaravanhoc commented:
Hey, thanks for your patience and help. I added a DEBUG_LSP macro for debug messages.
Everything should now be accounted for.

dekaravanhoc commented:
> It has been some time since you worked on this. @dekaravanhoc are you still interested in finishing this? Or can someone else pick up the work?

Hey, this was finished from my site 4 month ago. As stated in the last comment. All the remarks were resolved.

HolonProduction commented:
Haven't reread the whole conversation but the last thing I see are some minor style adjustments that are still missing. And the branch has conflicts so a rebase is needed. I won't be available for reviews for the next two weeks, but I'll give the PR another look after that.

dekaravanhoc commented:
Hey, :)

I rebased to master. If there is something open (style adjustments etc.), let me know. In April I worked through all your remarks as best as I could.
Also I worked with this for month now and had no issues. So at least from a technical standpoint it seems to be rather stable and working as intended.

Best regards

HolonProduction commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -403,3 +409,200 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+Node *SceneCache::get(const String &p_path) {
+	if (not cache.has(p_path)) {
+		DEBUG_LSP("Non loaded cache requested for:", p_path);
+		if (resource_request_queue[0] != p_path) {
+			resource_request_queue.insert(0, p_path);
+			_request_owner_scene_load_from_queue();
+		}
+		while (not cache.has(p_path)) {
+			_cache_current_requested_resource();
+			_request_owner_scene_load_from_queue();
+		}
```
This code does look wrong to me. I can't really test it since vscode always sends those openDocument signals way in advance, but some things to note:

- This code crashes with an empty resource queue, this can happen when the client did not send an open notification. This would be a client bug, but we should not crash because of it.
- It replaces the head of the request queue, this seems to work at the moment, but I'm very worried that it might lead to some invalid state in some edge case.
- The while loop at the end would get stuck if loading failed for some reason. This makes graceful error handling using the error macros very dangerous. While you don't use those at the moment, the regression potential is way to high IMO.

HolonProduction reviewed (commented):
Sorry for the delay, to be honest I'm having a hard time wrapping my head around this code. Like I can review the inner method code, and it looks mostly good aside from some minor stuff. But the control flow is really complicated to me, and I haven't really had the concentration to give it a full review.

However I suspect the current approach is not ready for big projects. Instances are cached based on script paths, not their own path. That means for big projects were a scene could contain a lot of scripts, this implementation would not reuse the instance between scripts. This might have quite a negative effect on memory usage.

I don't have a realistic project to verify this assumption, but I think we need to cache based on owner path, and cache the loaded owner for a given script separately to prevent this potential issue.

Also just some general ideas to make the control flow easier to understand:
- replace recursion with `while` (probably very subjective, so no requirement)
- From my understanding `owners_path_cache` is only ever populated for the current head of the resource request queue. Using a hash map is unnecessary. Converting it to a secondary queue that is only used for the head of the resource request queue, would make the loading process more clear IMO
- Usage of `ERR_FAIL_COND` at top of functions to make clear in what state the function is expected to be called. E.g. `_cache_current_requested_resource` has an early return is the queue is empty. This makes it hard for me to tell the context in which the method can be called. Realistically it does only make sense to call this method when the queue is not empty, so this should be an explicit error. `ERR_FAIL_COND(resource_request_queue.is_empty())`.

dekaravanhoc commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -34,6 +34,12 @@
+#define DEBUG_LSP(...) print_line("[ LSP -", __FILE__, ":", __LINE__, "-", __func__, "] -", ##__VA_ARGS__)
+#else
+#define DEBUG_LSP(...) ((void)0)
+#endif
```
Ah. Sorry. I understood this wrong. Changed it to its own definition. I kept the macro as I find it very convinient as it does not cluster the code with tons of ifdefs

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.cpp:
```diff
@@ -403,3 +409,200 @@ GDScriptLanguageProtocol::GDScriptLanguageProtocol() {
+				r_owners.push_back(p_dir->get_file_path(i));
+				break;
+			}
+		}
```
Changed it.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -35,12 +35,42 @@
+	void _cache_current_requested_resource();
+
+public:
+	bool has(const String &p_path);
```
removed.

dekaravanhoc commented on modules/gdscript/language_server/gdscript_language_protocol.h:
```diff
@@ -35,12 +35,42 @@
+public:
+	bool has(const String &p_path);
+	Node *get(const String &p_path);
+	Node *get_for_uri(const String &p_uri);
```
Cleaned everything up.

HolonProduction commented:
> I am not with you on this one. Feels like a grand over-optimization. Instances are afaik cheap and resources are shared. Instancing an already loaded scene happens instantly without any measurable impact on performance or memory. Even if you would instance 20 or 30 or 50 instances of the same scene. Would be pretty bad for godot, if that would heavily impact performance. But who opens 50 scripts (or even 20) in a session not even talking about a scene with 50 distinctive scripts?
> Or are we talking about different things here?

Instantiation can involve tool scripts and having user code involved at any place has me always a bit concerned. But I'm probably overthinking this. And you are definitely right that this PR is better than the status quo. So I guess I'll just shelve those concerns for now. If someone actually complains, it's something that can be fixed.

I will try to look through the control flow sometime this week. Hopefully I make it through this time :sweat_smile:
